packages/styled-components/src/utils/useCheckClassNameUsage.js

```diff
@@ -0,0 +1,78 @@
+// @flow
+import { useRef, useEffect, useCallback, type Ref, type ElementType } from 'react';
+import ReactDOM from 'react-dom';
+import getComponentName from './getComponentName';
+import { IS_BROWSER } from '../constants';
+import type { Target } from '../types';
+
+function useCombinedRefs<T: ElementType>(ref1: Ref<T>, ref2: Ref<T> | null) {
+  return useCallback(
+    (ref: any) => {
+      if (typeof ref1 === 'function') {
+        ref1(ref);
+      } else if (typeof ref1 === 'object' && ref1 !== null) {
+        // eslint-disable-next-line no-param-reassign
+        ref1.current = ref;
+      }
+
+      if (typeof ref2 === 'function') {
+        ref2(ref);
+      } else if (typeof ref2 === 'object' && ref2 !== null) {
+        // eslint-disable-next-line no-param-reassign
+        ref2.current = ref;
+      }
+    },
+    [ref1, ref2]
+  );
+}
+
+// this sounds like it should be a WeakSet instead
+const didWarnAboutClassNameUsage = new Set();
+function useDevelopmentCheckClassNameUsage(
+  ref: Ref<*>,
+  target: Target,
+  generatedClassName: string,
+  suppressClassNameWarning: boolean
+) {
+  const ownRefDefault: { current: any } = useRef(null);
+  const ownRef = suppressClassNameWarning ? null : ownRefDefault;
+  const refToUse = useCombinedRefs(ref, ownRef);
```

Comment:
As I said earlier, I tried implementing this by using refs to the child component, but this would end up passing refs even to pure function components.